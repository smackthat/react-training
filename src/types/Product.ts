export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
}

export interface ProductsState {
  products: Product[];
  loading: boolean;
  error: boolean;
}

export enum ProductCategory {
  MENS_CLOTHING = `men's clothing`,
  JEWELERY = `jewelery`,
  ELECTRONICS = `electronics`,
  WOMENS_CLOTHING = `women's clothing`,
}
