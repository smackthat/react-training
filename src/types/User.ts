export interface User {
  id: number;
  userName: string;
}

export interface CartItem {
  productId: number;
  title: string;
  quantity: number;
  sum: number;
}

export interface Cart {
  products: CartItem[];
}

export interface UserState {
  user: User;
  loading: boolean;
  error: UserErrorType[];
  cart?: Cart;
}

export enum UserErrorType {
  RESPONSE_ERROR,
  USER_NOT_FOUND,
  USER_NAME_EMPTY,
  PASSWORD_EMPTY,
}

export interface UserLogin {
  name: string;
  password: string;
}
