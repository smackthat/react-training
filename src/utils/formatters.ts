/** Currency formatter */
export function CurrencyFormatter(): Intl.NumberFormat {
  return new Intl.NumberFormat(undefined, {
    style: 'currency',
    currency: 'EUR',
  });
}
