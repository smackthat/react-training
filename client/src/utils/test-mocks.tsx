import { Product, ProductCategory } from 'types/Product';
import { CartItem, User, UserState } from 'types/User';
import { currencySum } from './helpers';

export const mockUser: User = {
  id: 2,
  userName: 'TEST',
};

export const mockLoggedInUserState: UserState = {
  user: mockUser,
  loading: false,
  error: null,
  cart: {
    items: [],
  },
};

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Shirt',
    description: 'A nice shirt',
    price: 9.99,
    category: ProductCategory.MENS_CLOTHING,
    image: '',
  },
  {
    id: 2,
    title: 'Skirt',
    description: 'A nice skirt',
    price: 19.99,
    category: ProductCategory.WOMENS_CLOTHING,
    image: '',
  },
  {
    id: 3,
    title: 'Phone X',
    description: 'A nice phone',
    price: 999.99,
    category: ProductCategory.ELECTRONICS,
    image: '',
  },
  {
    id: 4,
    title: 'Golden necklace',
    description: 'A nice necklace',
    price: 99.99,
    category: ProductCategory.JEWELERY,
    image: '',
  },
];

export const mockCartItems: CartItem[] = [
  {
    productId: mockProducts[0].id,
    title: mockProducts[0].title,
    quantity: 8,
    unitPrice: mockProducts[0].price,
    sum: currencySum(mockProducts[0].price, 8),
  },
  {
    productId: mockProducts[3].id,
    title: mockProducts[3].title,
    quantity: 1,
    unitPrice: mockProducts[3].price,
    sum: currencySum(mockProducts[3].price, 1),
  },
];
