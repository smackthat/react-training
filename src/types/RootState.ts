import { User } from './User';
// [IMPORT NEW CONTAINERSTATE ABOVE] < Needed for generating containers seamlessly

/* 
  Because the redux-injectors injects your reducers asynchronously somewhere in your code
  You have to declare them here manually
*/
export interface RootState {
  user: User;
  // [INSERT NEW REDUCER KEY ABOVE] < Needed for generating containers seamlessly
}
