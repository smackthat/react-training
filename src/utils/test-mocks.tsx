import { Product, ProductCategory } from 'types/Product';
import { User, UserState } from 'types/User';

export const mockUser: User = {
  id: 2,
  userName: 'TEST',
};

export const mockLoggedInUserState: UserState = {
  user: mockUser,
  loading: false,
  error: null,
  cart: {
    items: [],
  },
};

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Shirt',
    description: 'A nice shirt',
    price: 9.99,
    category: ProductCategory.MENS_CLOTHING,
    image: '',
  },
  {
    id: 2,
    title: 'Skirt',
    description: 'A nice skirt',
    price: 19.99,
    category: ProductCategory.WOMENS_CLOTHING,
    image: '',
  },
  {
    id: 3,
    title: 'Phone X',
    description: 'A nice phone',
    price: 999.99,
    category: ProductCategory.ELECTRONICS,
    image: '',
  },
  {
    id: 1,
    title: 'Golden necklace',
    description: 'A nice necklace',
    price: 99.99,
    category: ProductCategory.JEWELERY,
    image: '',
  },
];
