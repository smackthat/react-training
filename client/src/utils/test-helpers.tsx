import { I18nextProvider } from 'react-i18next';
import React, { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import i18nForTests from 'locales/i18nForTests';
import { configureAppStore } from 'store/configureStore';
import { RootState } from 'types/RootState';
import { EnhancedStore } from '@reduxjs/toolkit';
import { HelmetProvider } from 'react-helmet-async';
import { InjectedReducersType } from './types/injector-typings';
import { userReducer } from 'app/pages/LoginPage/slice';
import { productsReducer } from 'app/pages/HomePage/slice';
import { i18n } from 'locales/i18n';
import { mockProducts } from './test-mocks';

// See: https://testing-library.com/docs/react-testing-library/setup

export const defaultState: RootState = {
  productsState: {
    loading: false,
    error: false,
    products: mockProducts,
  },
  userState: {
    user: null,
    loading: false,
    error: null,
  },
};

const reducers: InjectedReducersType = {
  userState: userReducer,
  productsState: productsReducer,
};

async function customRender(
  ui: ReactElement,
  {
    initialState = defaultState,
    store = configureAppStore(initialState, reducers),
    ...options
  }: { initialState?: RootState; store?: EnhancedStore } = {},
) {
  const t = await i18n;

  return {
    result: render(
      <I18nextProvider i18n={i18nForTests}>
        <HelmetProvider>
          <Provider store={store}>{ui}</Provider>
        </HelmetProvider>
      </I18nextProvider>,
      options,
    ),
    store: store,
    t: t,
  };
}

export * from '@testing-library/react';
export { customRender as render };
