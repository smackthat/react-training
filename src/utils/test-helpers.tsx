import { I18nextProvider } from 'react-i18next';
import React, { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import i18n from 'locales/i18nForTests';
import { configureAppStore } from 'store/configureStore';
import { RootState } from 'types/RootState';
import { EnhancedStore } from '@reduxjs/toolkit';
import { HelmetProvider } from 'react-helmet-async';

// See: https://testing-library.com/docs/react-testing-library/setup

const defaultState: RootState = {
  productsState: {
    loading: false,
    error: false,
    products: [],
  },
  userState: {
    user: null,
    loading: false,
    error: null,
  },
};

function customRender(
  ui: ReactElement,
  {
    initialState = defaultState,
    store = configureAppStore(initialState),
    ...options
  }: { initialState?: RootState; store?: EnhancedStore } = {},
) {
  return {
    result: render(
      <I18nextProvider i18n={i18n}>
        <HelmetProvider>
          <Provider store={store}>{ui}</Provider>
        </HelmetProvider>
      </I18nextProvider>,
      options,
    ),
    store: store,
  };
}

export * from '@testing-library/react';
export { customRender as render };
