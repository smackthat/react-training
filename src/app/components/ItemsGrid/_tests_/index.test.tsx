import React from 'react';
import { defaultState, render } from 'utils/test-helpers';
import { fireEvent, screen, within } from '@testing-library/react';
import { ItemsGrid } from '..';
import { mockCartItems } from 'utils/test-mocks';
import { translations } from 'locales/translations';
import { RootState } from 'types';

const initState: RootState = {
  ...defaultState,
  userState: {
    ...defaultState.userState,
    cart: {
      items: mockCartItems,
    },
  },
};

const renderItemsGrid = (readOnly: boolean = false) =>
  render(
    <ItemsGrid
      items={initState.userState.cart.items}
      readOnly={readOnly}
    ></ItemsGrid>,
    {
      initialState: initState,
    },
  );

describe('ItemsGrid', () => {
  it('should match snapshot', async () => {
    const handler = await renderItemsGrid();
    expect(handler.result).toMatchSnapshot();
  });

  it('should update the totals sum correctly', async () => {
    const { t, store } = await renderItemsGrid();

    const row = screen.getByText(mockCartItems[0].title).closest('tr');
    const incrementButton = within(row).getByLabelText(
      t(translations.items.actions.increment) as string,
    );
    const decrementButton = within(row).getByLabelText(
      t(translations.items.actions.decrement) as string,
    );

    fireEvent.click(incrementButton);

    let updatedItems = (store.getState() as RootState).userState.cart.items;
    expect(updatedItems.map(i => i.quantity).reduce((a, b) => a + b)).toBe(10);

    fireEvent.click(decrementButton);

    updatedItems = (store.getState() as RootState).userState.cart.items;
    expect(updatedItems.map(i => i.quantity).reduce((a, b) => a + b)).toBe(9);
  });

  it('should not have item increment and decrement buttons when readonly', async () => {
    const { t } = await renderItemsGrid(true);

    const incrementButtons = screen.queryAllByLabelText(
      t(translations.items.actions.increment) as string,
    );
    const decrementButtons = screen.queryAllByLabelText(
      t(translations.items.actions.decrement) as string,
    );

    expect(incrementButtons).toHaveLength(0);
    expect(decrementButtons).toHaveLength(0);
  });
});
