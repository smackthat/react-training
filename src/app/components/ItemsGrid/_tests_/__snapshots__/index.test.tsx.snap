// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsGrid should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  overflow: auto;
}

.c0 Table {
  table-layout: fixed;
}

.c0 Table .totals td {
  font-weight: bold;
}

.c0 Table #deleteButton {
  margin-left: 1em;
}

<body>
    <div>
      <div
        class="MuiContainer-root c0 MuiContainer-maxWidthLg"
      >
        <table
          class="MuiTable-root MuiTable-stickyHeader"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
                  role="button"
                  tabindex="0"
                >
                  minicart.title
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiTableSortLabel-root"
                  role="button"
                  tabindex="0"
                >
                  minicart.quantity
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiTableSortLabel-root"
                  role="button"
                  tabindex="0"
                >
                  product.unitPrice
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiTableSortLabel-root"
                  role="button"
                  tabindex="0"
                >
                  minicart.sum
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              />
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                Golden necklace
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                €99.99
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                €99.99
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <button
                  aria-label="items.actions.increment"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="items.actions.decrement"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13H5v-2h14v2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                  id="deleteButton"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                Shirt
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                8
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                €9.99
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                €79.92
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <button
                  aria-label="items.actions.increment"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="items.actions.decrement"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13H5v-2h14v2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                  id="deleteButton"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root totals"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                minicart.total
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                9
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                €179.91
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root sc-AxjAm kQqKEu MuiContainer-maxWidthLg"
    >
      <table
        class="MuiTable-root MuiTable-stickyHeader"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
                role="button"
                tabindex="0"
              >
                minicart.title
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                tabindex="0"
              >
                minicart.quantity
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                tabindex="0"
              >
                product.unitPrice
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                tabindex="0"
              >
                minicart.sum
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Golden necklace
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              €99.99
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              €99.99
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <button
                aria-label="items.actions.increment"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="items.actions.decrement"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13H5v-2h14v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                id="deleteButton"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Shirt
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              8
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              €9.99
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              €79.92
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <button
                aria-label="items.actions.increment"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="items.actions.decrement"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13H5v-2h14v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                id="deleteButton"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root totals"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              minicart.total
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              9
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              €179.91
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
